metagraph/src/cli/query.cpp

```diff
@@ -279,224 +469,368 @@ construct_query_graph(const AnnotatedDBG &anno_graph,
+                                            && rev_contig[j + 1]
+                                            && full_dbg.has_single_outgoing(rev_contig[j]))) {
+                        call_hull_sequences(full_dbg, rev_path[j],
+                                            rev_contig.substr(j + 1, full_dbg.get_k() - 1),
+                                            callback, continue_traversal);
+                    }
+                }
+            }
 
-            size_t j = 0;
-            full_dbg.map_to_nodes(contig, [&](auto node_in_full) {
-                index_in_full_graph[path[j++]] = node_in_full;
-            });
-            assert(j == path.size());
+            #pragma omp critical
+            {
+                hull_contig_count += added_paths.size() + added_paths_rc.size();
+                for (auto&& pair : added_paths) {
+                    assert(pair.first.size() == pair.second.size() + full_dbg.get_k() - 1);
+                    contig_buffer.emplace_back(std::move(pair));
+                }
+                for (auto&& pair : added_paths_rc) {
+                    assert(full_dbg.is_canonical_mode()
+                        || pair.first.size() == pair.second.size() + full_dbg.get_k() - 1);
+                    rev_comp_contig_buffer.emplace_back(std::move(pair));
+                }
+            }
         }
 
-        logger->trace("[Query graph construction] Contigs mapped to graph in {} sec", timer.elapsed());
-        timer.reset();
+        assert((canonical && contig_buffer.size() == rev_comp_contig_buffer.size())
+                || (!canonical && rev_comp_contig_buffer.empty()));
+
+        for (auto&& pair : contig_buffer) {
+            graph_init->add_sequence(pair.first, [&](node_index) { ++num_added; });
+            contigs.emplace_back(std::move(pair));
+        }
+        for (auto&& pair : rev_comp_contig_buffer) {
+            graph_init->add_sequence(pair.first, [&](node_index) { ++num_added; });
+            rc_contigs.emplace_back(std::move(pair));
+        }
+
+        assert((canonical && contigs.size() == rc_contigs.size())
+                || (!canonical && rc_contigs.empty()));
+
+        logger->trace("[Query graph extension] Added {} nodes from {} contigs in {} sec",
+                      num_added, hull_contig_count, timer.elapsed());
     }
 
-    contigs = decltype(contigs)();
+    // merge rc_contigs and contigs
+    for (size_t i = 0; i < rc_contigs.size(); ++i) {
+        std::vector<node_index> &nodes = contigs[i].second;
+        std::vector<node_index> &dual_nodes = rc_contigs[i].second;
+        std::reverse(dual_nodes.begin(), dual_nodes.end());
+        assert(nodes.size() == dual_nodes.size());
+
+        for (size_t j = 0; j < nodes.size(); ++j) {
+            if (!nodes[j]) {
+                nodes[j] = dual_nodes[j];
+            } else if (dual_nodes[j]) {
+                nodes[j] = std::min(nodes[j], dual_nodes[j]);
```

Comment:
`std::min` won't work here for the hash ordered graph, since it uses the min of the k-mer encodings rather than the node index to determine canonical k-mers

Reply:
Yes, thanks.

I thought we would use a generic `map_to_annotated_nodes` moved to `AnnotatedDBG`, but it's easier to call the existing method for now.